c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +379,320 @@ Orphan<DynamicValue> JsonCodec::decode(
+    kj::Vector<char> decoded;
+
+    do {
+      auto stringValue = consumeWhile([](const char chr) {
+          return chr != '"' && chr != '\\';
+      });
+
+      decoded.addAll(stringValue);
+
+      if (nextChar() == '\\') {  // handle escapes.
+        advance();
+        switch(nextChar()) {
+          case '"' : decoded.add('"' ); advance(); break;
+          case '\\': decoded.add('\\'); advance(); break;
+          case '/' : decoded.add('/' ); advance(); break;
+          case 'b' : decoded.add('\b'); advance(); break;
+          case 'f' : decoded.add('\f'); advance(); break;
+          case 'n' : decoded.add('\n'); advance(); break;
+          case 'r' : decoded.add('\r'); advance(); break;
+          case 't' : decoded.add('\t'); advance(); break;
+          case 'u' :
+            advance();  // consume 'u'
+            unescapeAndAppend(kj::arrayPtr(remaining_.begin(), 4), decoded);
+            advance(4);
+            break;
+          default: KJ_FAIL_REQUIRE("Invalid escape in JSON string."); break;
+        }
+      }
+
+    } while(nextChar() != '"');
+
+    consume('"');
+    decoded.add('\0');
+
+    // TODO(perf): This copy can be eliminated, but I can't find the kj::wayToDoIt();
+    return kj::String(decoded.releaseAsArray());
+  }
+
+  kj::String consumeNumber() {
+    auto originalPos = remaining_.begin();
+
+    tryConsume('-');
+    if (!tryConsume('0')) {
+      consumeOne([](char c) { return '1' <= c && c <= '9'; });
+      consumeWhile([](char c) { return '0' <= c && c <= '9'; });
+    }
+
+    if (tryConsume('.')) {
+      consumeWhile([](char c) { return '0' <= c && c <= '9'; });
+    }
+
+    if (tryConsume('e') || tryConsume('E')) {
+      tryConsume('+') || tryConsume('-');
+      consumeWhile([](char c) { return '0' <= c && c <= '9'; });
+    }
+
+    KJ_REQUIRE(remaining_.begin() != originalPos, "Expected number in JSON input.");
+
+    kj::Vector<char> number;
+    number.addAll(originalPos, remaining_.begin());
+    number.add('\0');
+
+    return kj::String(number.releaseAsArray());
+  }
+
+  // TODO(someday): This "interface" is ugly, and won't work if/when surrogates are handled.
+  void unescapeAndAppend(kj::ArrayPtr<const char> hex, kj::Vector<char>& target) {
+    KJ_REQUIRE(hex.size() == 4);
+    int codePoint = 0;
+
+    for (int i = 0; i < 4; ++i) {
+      char c = hex[i];
+      codePoint <<= 4;
+
+      if ('0' <= c && c <= '9') {
+        codePoint |= c - '0';
+      } else if ('a' <= c && c <= 'f') {
+        codePoint |= c - 'a';
+      } else if ('A' <= c && c <= 'F') {
+        codePoint |= c - 'A';
+      } else {
+        KJ_FAIL_REQUIRE("Invalid hex digit in unicode escape.", c);
+      }
+    }
+
+    // TODO(soon): Support at least basic multi-lingual plane, ie ignore surrogates.
+    KJ_REQUIRE(codePoint < 128, "non-ASCII unicode escapes are not supported (yet!)");
+    target.add(0x7f & static_cast<char>(codePoint));
+  }
+
+private:
+  const size_t maxNestingDepth_;
```

Comment:
Hey, sorry to nitpick, but I just noticed you're doing Google-style trailing underscores on members. We don't use that style in KJ code -- could you remove them?